text/0000-simple-postfix-macros.md

```diff
@@ -0,0 +1,171 @@
+style of postfix macro, designed specifically for `await!` and for future cases
+like `try!` and `await!`, without blocking potential future extensions.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a macro using `macro_rules!`, you can include a first argument
+that uses a designator of `self` (typically `$self:self`). This must appear as
+the first argument, and outside any repetition. If the macro includes such a
+case, then Rust code may invoke that macro using the method-like syntax
+`expr.macro!(args)`. The Rust compiler will expand the macro into code that
+receives the evaluated `expr` as its first argument.
+
+```rust
+macro_rules! log_value {
+    ($self:self, $msg:expr) => ({
+        eprintln!("{}:{}: {}: {:?}", file!(), line!(), $msg, $self);
+        $self
+    })
+}
+
+fn value<T: std::fmt::Debug>(x: T) -> T {
+    println!("evaluated {:?}", x);
+    x
+}
+
+fn main() {
+    value("hello").log_value!("value").len().log_value!("len");
+}
+```
+
+This will print:
+
+```
+evaluated "hello"
+src/main.rs:14: value: "hello"
+src/main.rs:14: len: 5
+```
+
+Notice that `"hello"` only gets evaluated once, rather than once per reference
+to `$self`, and that the `file!` and `line!` macros refer to the locations of
+the invocations of `log_value!`.
+
+A macro that accepts multiple combinations of arguments may accept `$self` in
+some variations and not in others. For instance, `await!` could allow both of
+the following:
+
+```rust
+await!(some_future());
+some_other_future().await!().further_future_computation().await!();
+```
+
+This method-like syntax allows macros to cleanly integrate in a left-to-right
+method chain, while still making use of control flow and other features that
+only a macro can provide.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When expanding a postfix macro, the compiler will effectively create a
+temporary binding for the value of `$self`, and substitute that binding
+for each expansion of `$self`. This stands in contrast to other macro
+arguments, which get expanded into the macro body without evaluation. This
+change avoids any potential ambiguities regarding the scope of the `$self`
+argument and how much it leaves unevaluated, by evaluating it fully.
+
+The `await!` macro, whether defined in Rust code or built into the compiler,
+would effectively have the following two cases:
+
+```rust
+macro_rules! await {
+    ($e:expr) => ({
+        // ... Current body of await! ...
+    })
+    ($self:$self) => (
+        await!($self)
+    )
+}
+```
+
+Note that postfix macros cannot dispatch differently based on the type of the
+expression they're invoked on. This includes whether the expression has type
+`T`, `&T`, or `&mut T`. The internal binding the compiler creates for that
+expression will have that same type.
+
+Calling `stringify!` on `$self` will return `"$self"`.
```

Comment:
What would be the problem with stringifying the evaluated expression? Or the unevaluated one, for that matter.

Reply:
This is basically another limitation caused by pre-evaluation. `expr.dbg!()` doesn't work if you can't use `stringify`. You could make `stringify` magic but that sidesteps the actual issue.

Also, it's not just `stringify` -- it's *any* macro that expects an unevaluated expression.

```rust
macro_rules! is_ident {
    ($i:ident) => { true };
    ($e:expr) => { false }
}

macro_rules! log {
    ($self:self) => {{
        println!("{:?}", is_ident!($self));
        $self
    }}
}

42.log!();
```

What does this print? It seems quite surprising for it to print `true`, but impossible for it to print `false`.

Reply:
I suppose you could say that within the body of a postfix macro, `$self` is a totally opaque expression which only matches `:expr`.